[Feat] 랭킹 페이지 구현

mhsong95 opened the pull request:
### 🔨 작업 내용 설명
"동네 랭킹" 버튼을 눌러 지역별 랭킹을 볼 수 있는 기능 구현

### 📑 구현한 내용
- 랭킹 모달 레이아웃 및 스타일링
- 지역 선택 드롭다운 메뉴 상태관리 구현
- 지역별 랭킹 정보(평점 정보)를 불러오는 API 라우트 및 서비스 구현
- 랭킹 정보를 불러오고 가공해주는 클라이언트 컨트롤러 `rankController.ts` 구현

### 🚧 주의 사항
- 첫 진입시 "광역시 - 도" 단위로 랭킹이 표시됩니다. 
- 랭킹을 불러오는 API는 사실 평점 정보를 불러오는 API입니다. 순서대로 정렬하지 않고 지역별 평점 정보를 반환하며, 정렬은 클라이언트에서 합니다. 
- 평점이 `NaN`일 때(`count === 0`일 때) `StarRate`과 `BarRate`에서 예외처리가 필요합니다.  
- `.env` 템플릿에서 누락된 항목들을 추가했습니다. 

### 스크린 샷
![스크린샷(218)](https://user-images.githubusercontent.com/24454874/142325396-37e8da7a-65b4-4f78-9134-23db0676c288.png)
![스크린샷(219)](https://user-images.githubusercontent.com/24454874/142325421-e8bbbe1d-cf05-4b76-9279-1d2cc8393893.png)
![스크린샷(220)](https://user-images.githubusercontent.com/24454874/142325430-8bb5e9e4-325f-479d-9a39-feb80822d413.png)
![스크린샷(221)](https://user-images.githubusercontent.com/24454874/142325438-fc43f6fa-f2f6-464f-837e-28594800e131.png)
![스크린샷(222)](https://user-images.githubusercontent.com/24454874/142325448-ab10949a-b2c5-4a3e-a4be-060c15c7f866.png)

close #95

mhsong95 committed:
Fix: 로그 제거, 환경변수 추가, API 반환값 수정

mhsong95 committed:
Feat: #95 - 랭킹 모달 개별 항목 UI

- 랭킹 모달에서 보여질 개별 지역에 대한 평점 바(bar) 구현
- 클릭시 아래로 세부 평점이 펼쳐지도록 구현
- 원 모양 평점 UI 구현

mhsong95 committed:
Feat: #95 - 랭킹 모달 리스트 래퍼 구현

- Rankbar들을 감싸는 div 구현

mhsong95 committed:
Feat: #95 - 랭킹 모달 레이아웃 완료

- 제목, 지역 선택 버튼, 랭킹 리스트 배치 및 스타일링 완료

mhsong95 committed:
Feat: #95 - 랭킹 모달 상태관리 로직 구현

- 광역시·도 버튼과 시·군·구 버튼이 눌렸을 때 동작 방식 구현
- API 요청을 위한 스켈레톤 작성

mhsong95 committed:
Feat: #95 - 랭킹 API 구현

- `/rank` 라우트 컨트롤러 구현
- `services/rankService` 구현

mhsong95 committed:
Feat: #95 - 랭킹 fetch 구현

- "동네 랭킹" 메뉴를 클릭해 처음 진입하면 '광역시·도' 단위로 랭킹이 표시
- 두 개의 드롭다운 셀렉터를 사용하여 랭킹을 보고싶은 지역을 선택
- 첫 번째 셀렉터가 "광역시 ·  도"로 선택되어 있으면 두 번째 셀렉터는 비활성화

mhsong95 committed:
클라이언트 누락된 환경변수 추가

hongjw1938 commented on client/src/components/RankingModal/Selector/index.tsx:
```diff
@@ -0,0 +1,51 @@
+  const toggleDropdown = useCallback(
+    () => setDropdownOpen((prev) => !prev),
+    [],
+  );
```
상위의 toggleDetail은 useCallback을 쓰지 않았는데 여기에는 써주신 이유가 있을까요?

mhsong95 replied:
오히려 `toggleDetail`에서 `useCallback`을 빼뜨린 것 같네요. 함수들은 웬만하면 `useCallback`을 사용하려 하고 있습니다. 재렌더링될 때마다 함수가 생성되는 걸 방지하기 위해서요.

hongjw1938 reviewed (approved):
빠른 시간 내에 잘 구현해주셨습니다!

mhsong95 committed:
Feat: #95 - 지역 선택 드롭다운 메뉴 hover 스타일 추가

- 드롭다운 메뉴에 마우스 올려놓을 때 배경이 흐려지는 효과 추가
- 랭킹 리스트가 바뀔 때 레이블이 바뀌었음에도 랭크바가 펼쳐져있는 문제 해결
- useCallback, useMemo를 사용한 최적화
